src/main/cpp/faultinj/faultinj.cu

```diff
@@ -0,0 +1,453 @@
+                    cbInfo->functionName, cbInfo->symbolName);
+                return;
+            }
+        }
+        matchedFaultConfig = lookupConfig(globalControl.runtimeFaultConfigs, cbInfo->functionName, cbid);
+    }
+    PTHREAD_CALL(pthread_rwlock_unlock(&globalControl.configLock));
+
+    if (!matchedFaultConfig) {
+        return;
+    }
+
+    const int injectionType = (*matchedFaultConfig)
+        .get_optional<int>("injectionType")
+        .value_or(static_cast<int>(FI_RETURN_VALUE));
+
+    const int substituteReturnCode = (*matchedFaultConfig)
+        .get_optional<int>("substituteReturnCode")
+        .value_or(static_cast<int>(CUDA_SUCCESS));
+
+    const int injectionProbability = (*matchedFaultConfig)
+        .get_optional<int>("percent")
+        .value_or(0);
+
+    spdlog::trace("considered config domain={} function={} injectionType={} probability={}",
+        domain, cbInfo->functionName, injectionType, injectionProbability);
+    if (injectionProbability < 100) {
+        if (injectionProbability <= 0) {
+            return;
+        }
+        const int rand10000 = rand() % 10000;
+        const int skipThreshold = injectionProbability * 10000 / 100;
+        spdlog::trace("rand1000={} skipThreshold={}", rand10000, skipThreshold);
+        if (rand10000 >= skipThreshold) {
+            return;
+        }
+        spdlog::debug(
+            "matched config based on rand10000={} skipThreshold={} "
+            "domain={} function={} injectionType={} probability={}",
+            rand10000, skipThreshold, domain,
+            cbInfo->functionName, injectionType, injectionProbability);
+    } else {
+        spdlog::debug("matched 100% config domain={} function={} injectionType={} probability={}",
+            domain, cbInfo->functionName, injectionType, injectionProbability);
+    }
+
+    switch (injectionType)
+    {
+    case FI_TRAP:
+        faultInjectorKernelTrap<<<1,1>>>();
+        break;
+
+    case FI_ASSERT:
+        faultInjectorKernelAssert<<<1,1>>>();
+        break;
+
+    case FI_RETURN_VALUE:
+        if (domain == CUPTI_CB_DOMAIN_DRIVER_API) {
+            CUresult *cuResPtr = static_cast<CUresult*>(cbInfo->functionReturnValue);
+            *cuResPtr = static_cast<CUresult>(substituteReturnCode);
+        } else if (domain == CUPTI_CB_DOMAIN_RUNTIME_API) {
+            cudaError_t *cudaErrPtr = static_cast<cudaError_t*>(cbInfo->functionReturnValue);
+            spdlog::error("updating runtime return value DOES NOT WORK, use trap or assert");
+            *cudaErrPtr = static_cast<cudaError_t>(substituteReturnCode);
+            break;
+        }
+
+    default:
+        break;
+    }
+}
+
+/**
+ * cuInit hook entry point
+ */
+int STDCALL
+InitializeInjection(void) {
+    spdlog::info("InitializeInjection");
+    if (globalControl.initialized) {
+        return 1;
+    }
+    // Init globalControl
+    globalControlInit();
+    globalControl.initialized = 1;
```

Comment:
Nitpick:

I wonder if we might streamline the initialization/management of `globalControl` state from one place. It looks like `globalControl.initialized` is first set to `0`, then changed to `1` *after* (and *outside*) `globalControlInit()`. 

Would there be value in using a `static` constructor to initialize this singleton? It might obviate the need for `initialized`:

```c++
injGlobalControl& STDCALL
￼globalControl(void) {
￼    spdlog::info("InitializeInjection");
     
     static injGlobalControl instance;
     // Guaranteed to be initialized, thread-safely.

     // Move PTRHEAD_CALL(), `cuptiInitialize()`, etc. into body of constructor.

     return &instance;
}
```

`instance` would be initialized with thread-safety, on first call.

Reply:
I think there are many opportunities to make the initial code look more C++ than C. Is it OK if we handle as a separate epic with various suggestions you are making

Reply:
Yes, sir. Of course. None of what I'm pointing out warrants blocking this.